File: src/utils/chat/permissions.ts
export const PERMISSIONS = {
  READ_MESSAGE: 1n << 0n,
  SEND_MESSAGE: 1n << 1n,
  ATTACH_FILE_MESSAGE: 1n << 2n,
  ADD_REACTION_MESSAGE: 1n << 3n,
  PIN_MESSAGE: 1n << 4n,
  CONNECT_VOCAL: 1n << 5n,
  SPEAK_VOCAL: 1n << 6n,
  VIDEO_VOCAL: 1n << 7n,
  SCREENSHARE_VOCAL: 1n << 8n,
  WATCHTOGETHER_VOCAL: 1n << 9n,
  MUTE_MEMBER_VOCAL: 1n << 10n,
  DEAFEN_MEMBER_VOCAL: 1n << 11n,
  DISABLE_VIDEO_VOCAL: 1n << 12n,
  MOVE_MEMBER_VOCAL: 1n << 13n,
  DELETE_MESSAGE: 1n << 14n,
  MUTE_MEMBER: 1n << 15n,
  KICK_MEMBER: 1n << 16n,
  BAN_MEMBER: 1n << 17n,
  VIEW_CHAT_LOG: 1n << 18n,
  MANAGE_INVITE: 1n << 19n,
  MANAGE_SUB: 1n << 20n,
  ASSIGN_ROLE: 1n << 21n,
  MANAGE_ROLE: 1n << 22n,
  MANAGE_CHAT: 1n << 23n,
} as const;

export const DEFAULT_ROLES = {
  OWNER: { id: 0 },
  ADMIN: { id: 1 },
  USER: { id: 2 },
} as const;

export const SUB_WRITE_ALLOWED_ROLES: Record<string, Set<number>> = {
  ANNOUNCE: new Set([DEFAULT_ROLES.OWNER.id, DEFAULT_ROLES.ADMIN.id]),
};

export function hasPermission(
  memberRoles: Array<{ id: number; permission: string; level: number }>,
  requiredPermission: bigint,
  subType?: string,
): boolean {
  if (!memberRoles || memberRoles.length === 0) return false;

  const allowedRoles = subType ? SUB_WRITE_ALLOWED_ROLES[subType] : null;

  let permissions = 0n;

  for (const role of memberRoles) {
    let rolePerms = BigInt(role.permission);

    if (allowedRoles && !allowedRoles.has(role.id)) {
      rolePerms = rolePerms & ~PERMISSIONS.SEND_MESSAGE;
      rolePerms = rolePerms & ~PERMISSIONS.ATTACH_FILE_MESSAGE;
    }

    permissions |= rolePerms;
  }

  return (permissions & requiredPermission) === requiredPermission;
}

export function getEffectiveLevel(
  memberRoles: Array<{ level: number }>,
): number {
  let level = 0;
  if (!memberRoles) return level;
  for (const role of memberRoles) {
    if (role.level > level) {
      level = role.level;
    }
  }
  return level;
}

File: lib/core/chat/permissions.dart
/// Chat role permissions bitmask definitions.
class ChatPermissions {
  static final BigInt readMessage = BigInt.from(1) << 0;
  static final BigInt sendMessage = BigInt.from(1) << 1;
  static final BigInt attachFileMessage = BigInt.from(1) << 2;
  static final BigInt addReactionMessage = BigInt.from(1) << 3;
  static final BigInt pinMessage = BigInt.from(1) << 4;
  static final BigInt connectVocal = BigInt.from(1) << 5;
  static final BigInt speakVocal = BigInt.from(1) << 6;
  static final BigInt videoVocal = BigInt.from(1) << 7;
  static final BigInt screenshareVocal = BigInt.from(1) << 8;
  static final BigInt watchtogetherVocal = BigInt.from(1) << 9;
  static final BigInt muteMemberVocal = BigInt.from(1) << 10;
  static final BigInt deafenMemberVocal = BigInt.from(1) << 11;
  static final BigInt disableVideoVocal = BigInt.from(1) << 12;
  static final BigInt moveMemberVocal = BigInt.from(1) << 13;
  static final BigInt deleteMessage = BigInt.from(1) << 14;
  static final BigInt muteMember = BigInt.from(1) << 15;
  static final BigInt kickMember = BigInt.from(1) << 16;
  static final BigInt banMember = BigInt.from(1) << 17;
  static final BigInt viewChatLog = BigInt.from(1) << 18;
  static final BigInt manageInvite = BigInt.from(1) << 19;
  static final BigInt manageSub = BigInt.from(1) << 20;
  static final BigInt assignRole = BigInt.from(1) << 21;
  static final BigInt manageRole = BigInt.from(1) << 22;
  static final BigInt manageChat = BigInt.from(1) << 23;
}

/// Standard predefined role IDs.
class DefaultRoles {
  static const int owner = 0;
  static const int admin = 1;
  static const int user = 2;
}

/// Sub-channel types with specific role write permissions.
final Map<String, Set<int>> subWriteAllowedRoles = {
  'ANNOUNCE': {DefaultRoles.owner, DefaultRoles.admin},
};

/// Evaluates if a member with given [memberRoles] possesses the [requiredPermission].
///
/// If [subType] is specified (e.g. `'ANNOUNCE'`), restrictions on sending messages
/// or attaching files are enforced based on [subWriteAllowedRoles].
bool hasPermission(
  List<dynamic> memberRoles,
  BigInt requiredPermission, [
  String? subType,
]) {
  if (memberRoles.isEmpty) return false;

  final allowedRoles = subType != null ? subWriteAllowedRoles[subType] : null;
  var permissions = BigInt.zero;

  for (final roleRaw in memberRoles) {
    if (roleRaw is! Map) continue;
    final roleId = roleRaw['id'] is num
        ? (roleRaw['id'] as num).toInt()
        : int.tryParse(roleRaw['id']?.toString() ?? '0') ?? 0;

    var rolePerms = BigInt.zero;
    final permVal = roleRaw['permission'];
    if (permVal is num) {
      rolePerms = BigInt.from(permVal);
    } else if (permVal is String) {
      rolePerms = BigInt.tryParse(permVal) ?? BigInt.zero;
    } else if (permVal is BigInt) {
      rolePerms = permVal;
    }

    if (allowedRoles != null && !allowedRoles.contains(roleId)) {
      rolePerms = rolePerms & ~ChatPermissions.sendMessage;
      rolePerms = rolePerms & ~ChatPermissions.attachFileMessage;
    }

    permissions |= rolePerms;
  }

  return (permissions & requiredPermission) == requiredPermission;
}

/// Returns the highest role level among the member's roles.
int getEffectiveLevel(List<dynamic> memberRoles) {
  var level = 0;
  for (final roleRaw in memberRoles) {
    if (roleRaw is! Map) continue;
    final roleLevel = roleRaw['level'] is num
        ? (roleRaw['level'] as num).toInt()
        : int.tryParse(roleRaw['level']?.toString() ?? '0') ?? 0;

    if (roleLevel > level) {
      level = roleLevel;
    }
  }
  return level;
}
